### ISSUE TITLE
[API Proposal]: StringValues implicit conversion to string?[]? causes confusing error with Logging methods

### ISSUE THREAD

**Issue by Jay-Madden** [dotnet/runtime#78241]
## Background and Motivation
Attempting to log a header value from Headers.TryGetValue in c#11 wont compile due to StringValues

This causes a rather nasty error with logging. Take this code 
```cs
if (_httpContextAccessor.HttpContext!.Request.Headers.TryGetValue("Origin", out var origin))
{
    _logger.LogInformation("Site Login Request Initialized from Url: {Origin}", origin);
}
```

Origin will error here with error `Possible null reference argument for parameter 'args' in 'Microsoft.Extensions.Logging.LoggerExtensions`

Because of the implicit conversion to string?[]? here due to logging methods taking a params object[ as an argument: https://github.com/dotnet/runtime/blob/2b1f57e08b26879e4fa99670df078b77c848d9e0/src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs#L81

 but that cause is incredibly non obvious. 

### API Proposal

Not 100% sure what a good design for this is, but the current ergonomics are incredibly unpleasant. Having a struct that can represent null but implicitly converts to a ref type when targeting object/string is quite annoying. 

Perhaps an implicit conversion to an empty array for conversions to a non nullable object array.
```cs
        public static implicit operator object?[] (StringValues value)
        {
            return value.GetArrayValue() ?? new object[]{};
        }
``` 

Just throwing something out there, this has its own problems

### API Usage

Same as current, just different overload selected what passed to logging methods

### Alternative Designs

A named method could be a decent compromise here, while still a breaking change at least its visible.
```cs

public object?[] AsStringArrayOrEmpty(StringValues value)
{
    value.GetArrayValue() ?? Array.Empty<Object>();
}
```

### Risks

Changing representations from null to empty array for specific conversions has the obvious implications of being a breaking change where the implicit null is a relied upon implementation detail.

**Comment by Jay-Madden** [comment]
Roslyn issue opened here https://github.com/dotnet/roslyn/issues/65369

**Comment by RikkiGibson** [comment]
Many conversion operators use NotNullIfNotNullAttribute to indicate that the conversion result is always not-null if the conversion operand is not-null. Is that something that should be used here?

**Comment by rosebyte** [comment]
> edit: oh, StringValues is a struct type, I think there's nothing to improve here in terms of annotations :)

Yes, it is a struct, but the issue is that it implicitly converts to string?[]?

@Jay-Madden, I don't think StringValues should be directly added as the generic type for the Log method. The biggest issue isn't the nullability but the type itself - if there would be two header values, you likely get runtime exception that the number of arguments doesn't match the template. Please consider calling .ToString() on the values, or making a string variable to cast it to string first.

**Comment by cincuranet** [comment]
The implicit conversion to `string?[]?` is valuable and I don't think we should change it/remove it. The fact that the compiler changes that for `LogXxx` and not the `string?` is (un)lucky coincidence. I agree with @rosebyte that explicitly calling `ToString()` is good enough solution.
